### ISSUE TITLE
[rewrite] Suggestion: Rename data to body

### ISSUE THREAD

**Issue by gilbert** [MithrilJS/mithril.js#1138]
The term `body` is more descriptive than `data`, and is (I believe) the official term in HTTP verbiage. Personally, I've never liked how the non-descriptive `data` key had two different functions in jQuery (and consequently, made it harder to teach).

For GET requests, it might be appropriate to make `query` the key for query parameters. What do you think?

**Comment by isiahmeadows** [comment]
1. You're correct in that HTTP uses `body` to refer to the data sent with the request.
2. I'd prefer those to be synonyms at least short term (we can deprecate/remove `data` later, but let's leave it at that *first*).

**Comment by lhorie** [comment]
I like this idea

**Comment by mindeavor** [comment]
@isiahmeadows I see, yes, it makes sense to leave `data` as deprecated to make the migration from 0.2.x easier. What do you mean by synonyms?

For the record, html5 `fetch` also uses `body` as its key.

**Comment by tivac** [comment]
@mindeavor He means that as part of the transition both `data` **and** `body` should be acceptable, preferably with `data` logging some sort of deprecation warning.

**Comment by mindeavor** [comment]
Ah, I ask because I wouldn't want `body` to behave like the current `data` for `GET` requests, so, not a synonym...

**Comment by isiahmeadows** [comment]
@mindeavor 

> Ah, I ask because I wouldn't want body to behave like the current data for GET requests, so, not a synonym...

You've lost me...what do you mean by `params` (I'm guessing you meant this) *not* being a synonym of `data` for `GET` requests? What would be different other than name?

**Comment by lhorie** [comment]
my two cents is to deprecate `data` in favor of `query` and `body`, which should hopefully be immediately familiar to anyone who's used express/hapi/etc

**Comment by lhorie** [comment]
on second thought, `data` also handles `params`, so it may be worth keeping `data` as a "I-dont-care-how-stuff-translates-to-http" mechanism

**Comment by mindeavor** [comment]
@isiahmeadows I guess I didn't know what you meant by "those" when you said "I'd prefer those", which is why I asked for clarification.

@lhorie By "handles params", are referring to the [Dynamic URLs](https://github.com/lhorie/mithril.js/blob/rewrite/docs/request.md#dynamic-urls) feature?

**Comment by isiahmeadows** [comment]
@mindeavor (re: `params`) I presume so, since `GET` arguments generally involve URL parameters, not HTTP request bodies. I think he's referring to the fact `GET` query parameters in URLs and `POST` request bodies both represent the same abstraction: parameters for a function.

If you squint hard enough, these will seem awfully similar:

```js
const API = "https://api.stackexchange.com/2.2"
const buildQueryString = require("mithril/querystring/build")

// Get a list of answers
let answers = fetch(`${API}/answers/123456%3B238427%3B192528%3B324925?order=desc&sort=activity&site=stackoverflow`)
let answers = m.request({
  method: "GET",
  url: `${API}/answers/:ids`,
  params: {
    ids: [123456, 238427, 192528, 324925],
    order: "desc",
    sort: "activity",
    site: "stackoverflow",
  },
})
let answers = SE.Answer.get({
  ids: [123456, 238427, 192528, 324925],
  order: "desc",
  sort: "activity",
  site: "stackoverflow",
})

// Accept an answer
let result = fetch(`${API}/answers/123456/accept`, {
  method: "POST",
  body: "key=somekey&access_token=sometoken&site=stackoverflow",
  headers: {"Content-Type": "application/x-www-form-urlencoded"},
})
let result = m.request({
  method: "POST",
  url: `${API}/answers/:id/accept`,
  config: xhr => {
    xhr.setRequestHeader("Content-Type", "application/x-www-form-urlencoded")
    return xhr
  },
  serialize: buildQueryString,
  body: {
    id: 123456,
    key: "somekey",
    access_token: "sometoken",
    site: "stackoverflow",
  },
})
let result = SE.Answer.accept({
  id: 123456,
  key: "somekey",
  access_token: "sometoken",
  site: "stackoverflow",
})
```

(One of my biggest annoyances about requests: they're so freaking verbose for something that is so conceptually simple at a high level.)

**Comment by mindeavor** [comment]
I agree they're similar in many cases, but in others they're not; you might have both a request body and parameters in one request.

An HTTP request may be verbose, but it's also very clear, and easy to abstract over :)

**Comment by ikki3355** [comment]
@lhorie  "so it may be worth keeping data as a "I-dont-care-how-stuff-translates-to-http" mechanism"
I would prefer only having one name for one thing (eventually). It will make things simple and cause less confusion (less documentation and easy to teach).
people work with mithril should have some relevant background on HTTP.

I am not sure about the "params" for GET request. It should be part of the URL (query string).

**Comment by isiahmeadows** [comment]
Automatic query string building for GET requests is useful in my opinion,
though.

**Comment by ikki3355** [comment]
In my option, URL is a URL, body is a body. storing URL data in the "body" field will be confusing to people who are familiar with GET urls (I can hardly imagine any web developer who doesn't know GET urls). 

have read through the thread again.
isiahmeadows mentioned "automatic query string building", mindeavor posted a link on "Dynamic URL" feature.

I think we are complicating the implementation by treating some simple utility functions as fancy features of m.request().

```
var user = {id: 1};

m.request({
    method: "GET",
    url: "/api/v1/users/:id",
    data: user,
})
```
I don't see the above is more convient than
```
var user = {id: 1};

m.request({
    method: "GET",
    url: "/api/v1/users/" + user .id,
    // data: user  , // don't need data parameter for a GET request
})
```

for the auto params building, I think what we need is some convenient utility functions, `buildQueryString()` should be one of them. Simple and straightforward as below:
```
url: "/api/v1/users/" + user.id + m.buildQueryString(params)
```

**Comment by mindeavor** [comment]
@longztian I agree with your overall point. However, I think the case for `/api/v1/users/:id` is for POST request convenience, rather than GET. For example:

```js
function createInvite (attrs) {
  return m.request({
    method: 'POST',
    url: '/events/:event_id/invites',
    body: attrs,
  })
}

var invite = { name: 'My Event', event_id: 10, description: '...' }
createInvite(invite)
```

**Comment by longztian** [comment]
@mindeavor do you mean, if the event_id appears in the URL, it will get removed from "body"? That would make sense, but need some extra work when stringifying the attrs object.

otherwise, if we are OK with posting an extra field (event_id, server can ignore it), It seems no difference between your code and the below:
```javascript
function createInvite (invite) {
  return m.request({
    method: 'POST',
    url: "/events/" + invite.event_id + "/invite",
    body: invite,
  })
}

var invite = { name: 'My Event', event_id: 10, description: '...' }
createInvite(invite)
```

I am thinking if the POST URL is "/api/v1/invite" or "/api/v1/event/invite", we will need the event_id in the body, but won't need it in the url.

Anyway, my option is we don't need to build the URL automatically and internally for programmers, they may build it however/wherever they want.
just my 2 cents :)

**Comment by hugufc** [comment]
In that case string interpolation aka "template strings" comes in good way..

**Comment by longztian** [comment]
@hugufc do you mean "template literals"? I think it is neat too
```javascript
url: `events/${invite.event_id}/invite"
```
versus 
```javascript
url: "/events/" + invite.event_id + "/invite"
```

I prefer the pure javascript feature. not sure about creating our own templating rules (a hidden rule that the parameter name is the property name of the "data" object). This will steep the learning curve of mithril.

**Comment by hugufc** [comment]
@longztian Oh.. I didn't notice that they have changed the name in spec https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Template_literals. Thanks.

BTW. I agree in the use of the present "dynamic url" because it is defined in this same style in routes.

**Comment by isiahmeadows** [comment]
@hugufc @longztian I'll note that template literals won't escape things for you, but [there's a tag for that](https://www.npmjs.com/package/url-escape-tag). And this adds to the list of reasons why I feel `body`  and `query` should be *synonyms* for `data`. (BTW, the proposal included both `query` for mainly `GET`/`PUT`/`DELETE` requests, and `body` mainly for others)

**Comment by longztian** [comment]
@isiahmeadows the url-escape-tag does not look good to me. :)
I agree that we can rename the 'data' to 'body'.
'query' is part of URL so we need a utility function to append it to the URL. m.request shouldn't build the URL implicitly.

Again, I think we are complicating mithril core by adding utility functions as fancy features of m.request(). Moving the [Dynamic URLs](https://github.com/lhorie/mithril.js/blob/rewrite/docs/request.md#dynamic-urls) to a separated (and optional) utility function will make the core of Mithril (and the documentation) clean and simple.

So how about a `m.url()` utility function for assembling URLs?

**Comment by isiahmeadows** [comment]
@lhorie I'm kind of unsure how to approach @longztian's suggestion...I feel you could help answer that a bit better than I could.

@longztian Mind filing a separate issue for that suggestion. It's mostly orthogonal to the main topic here, and I'm hoping to pull this conversation back to the original proposal of making `body` and `query` aliases of `data` for `m.request` and `m.jsonp`.

**Comment by isiahmeadows** [comment]
Closing, since this issue/PR has not been updated in over a month, and typically, issues inactive for that long do not usually produce any action. If you feel something in Mithril needs added or changed, please file a new issue/PR.

**Comment by isiahmeadows** [comment]
Reopening because I'm looking to propose this to make it to v2, in the form of `params` for query parameters and `body` for the raw body.
